src/coreclr/jit/importer.cpp

```diff
@@ -5970,6 +5972,85 @@ bool Compiler::impBlockIsInALoop(BasicBlock* block)
            block->HasFlag(BBF_BACKWARD_JUMP);
 }
 
+//------------------------------------------------------------------------
+// impMatchAwaitPattern: check if a method call starts an Await pattern
+//                       that can be optimized for runtime async
+//
+// Arguments:
+//   codeAddr - IL after call[virt]
+//   codeEndp - End of IL code stream
+//   configVal - [out] set to 0 or 1, accordingly, if we saw ConfigureAwait(0|1)
+//
+// Returns:
+//    true if this is an Await that we can optimize
+//
+bool Compiler::impMatchAwaitPattern(const BYTE* codeAddr, const BYTE* codeEndp, int* configVal)
+{
```

Comment:
Here we detect the pattern where one Async method awaits another and turn it into async call. 
```cs
T result = AwaitHelpers.Await<T>(AsyncMehtod());

becomes 

// call Async variant, skip wrapping/unwrapping the result through `Task<T>`
T result = async_call RtAsyncMehtod(); 
```

(added a comment to reference from another PR)